python-package/lightgbm/sklearn.py

```diff
@@ -945,11 +945,12 @@ def fit(self, X, y,
             sample_weight=None, init_score=None, group=None,
             eval_set=None, eval_names=None, eval_sample_weight=None,
             eval_init_score=None, eval_group=None, eval_metric=None,
-            eval_at=[1, 2, 3, 4, 5], early_stopping_rounds=None, verbose=True,
+            eval_at=None, early_stopping_rounds=None, verbose=True,
             feature_name='auto', categorical_feature='auto',
             callbacks=None, init_model=None):
         """Docstring is inherited from the LGBMModel."""
         # check group data
+        eval_at = eval_at or [1, 2, 3, 4, 5]
```

Comment:
scikit-learn doesn't like when we modify `fit` arguments: https://github.com/microsoft/LightGBM/pull/3254#discussion_r483997893.

I believe changing default value from `list` to `tuple` (`[1, 2, 3, 4, 5] -> (1, 2, 3, 4, 5)`) will be enough and will not break anything.

Reply:
in this case, the behavior is identical. Just the way we achieve it is different. I followed that comment and then the PR it linked to and I'm not sure that the change I'm proposing would actually break anything.

I'm nervous that changing data types will break something, because now you have the default as a tuple and users probably passing in values as a list

Reply:
> I'm nervous that changing data types will break something, because now you have the default as a tuple and users probably passing in values as a list

That `pylint` warning you are addressing in this PR is about dangerous **default** values. There should be nothing wrong with that users pass lists intentionally. 

I don't really like you approach because with it default value is implicit: documentation says it is `1,2,3,4,5` but actually it is `None`. It may be misleading. 
Also, scikit-learn prefers everything to be explicit. It is not allowed to change `init` args or even have `kwargs`, for example. So I'm pretty sure it is better to not have implicit argument overwriting for avoiding any possible (future) issues.

Reply:
ok. To be honest I still don't understand the scikit-learn thing but I believe you.

>  There should be nothing wrong with that users pass lists intentionally.

Using a tuple as the argument means that if we have any code anywhere down the stack that is expecting specifically a list and not a tuple (for example, something that does `eval_at.append()`), it will break. But let me try and see what the tests say.

Reply:
sorry I forgot about this one! Just updated to a tuple. Hopefully the tests will still pass. I searched around with `git grep eval_at` and didn't find any list-specific code so I think it should work.